Update to use IFB, HFSC and FQ_CODEL

drwyrm opened the pull request:
* Uses just one IFB instead of 2 IMQs
* Changes config option to just "ifb" instead of "(down|up)load_ifb"
* Uses tc u32 filter instead of fwmark

drwyrm commented:
Can you please review this before i go through merge hell to update the patch to your master branch? Thanks!

PS: If you are going to test this on OpenWRT, then you probably want the patch from https://bugzilla.kernel.org/show_bug.cgi?id=84661 to get correct output from tc filter show

mwarning commented:
Thanks for the contribution!
The code looks good on first look and compiles. But I do not know  IFB.
I would like to merge it, since IMQ is not supported anymore.
@lynxis do you have any thought on this issue?

@drwyrm: As it looks, limiting the download is not possible anymore. Do I see that right?
Feel free to continue to fix the merge conflict.

bluewavenet commented:
It is good to see work being done on this. However is there any advantage in using IFB over SQM?
I have been successfully using SQM in a production environment with NDS. Binding to br-lan and using fq_codel qdisc and simple.qos gives extremely good results. With a 35mb/s FTTC we can have numerous clients streaming video with no noticeable drop in performance whereas previously, just one "streamer" would grind it almost to a halt. No integration with NDS is required. The production systems are running on APU2C2's (ie AMD 1GHz quad core with 2GB ram) so cpu power is not an issue, however I have also tested on a Ubiquiti AirGateway with similar results although admittedly not quite the same load. @drwyrm , what testing have you done so far?

drwyrm commented:
@mwarning It is still supported, it just doesn't need an intermediate queue device anymore. Download (It's actually upload/egress from router's point of view) is supported directly by br-lan/wlan?(-?) interfaces, and the intermediate interface is only needed to support user's upload (router's download/ingress) traffic.

You can still limit the total bandwidth for the interface and the bandwidth for each user individually using the binvoucher output, both interface and client limits were tested and work for both egress and ingress.

@bluewavenet I did no testing for compatibiliy issues with openwrt's sqm-scripts, I discarded sqm-scripts and went back to using qos-scripts because sqm-scripts added 4-5ms of latency to each interface while idle in relation to what i got with good old qos-scripts.

I'm using NDS with qos-scripts and mwan3, and the only compatibility issues I had is that i had to change qos-scripts script to increase the "numifbs" kmod option, and i have to set all three fw_mark_* (the default conflicts with mwan3) as well as the ifb option in nds config file (the default conflicts with qos-scripts).

Of course, qos-scripts is configured to handle the wan interface, and the nds-vlan is being handled by nds itself, I previously used qos-scripts for the wifi/lan interface too, but using NDS and being able to set limits per voucher/user/mac in adittion to a global limit for the interface like in qos-scripts is a lot more convenient. The rules I'm using in nds are based on qos-scripts, they still need tweaking, but already perform as good as qos-scripts itself (i have to keep it simpler than qos-scripts, because our ruleset grows with each authorized user).

On your setup, you will probably benefit a lot from moving from sqm-scripts to per-client QOS control with NDS (and keep sqm-scripts for your wan), as it will allow you to distribute bandwidth fairly for different clients instead of just streams (like fq_codel does), while keeping all the benefits of sqm-scripts since you can still use it on your wan interface (most likelly your bottleneck). I would be able to estimate this more acuratelly if you provide me your "tc -s qdisc" output so I know which rules are actually being triggered for your use case.

bluewavenet commented:
I am not saying I have a problem, on the contrary it is running incredibly well. I too am running mwan3, in my case with two vdsl internet feeds at around 36Mb/s each. SQM is bound to br-lan interface.
The question I am asking is if there is an advantage to running IFB over SQM? I suspect your latency was an issue with lack of cpu power, particularly on a wifi interface. Do we need client based control rather than connection based? Surely connection based is fairer?

drwyrm commented:
> The question I am asking is if there is an advantage to running IFB over SQM?
Yes, client based qos instead of just flow based shared for all users.

> Do we need client based control rather than connection based? Surely connection based is fairer?
This isn't an "one or the other" decision. You can have both if you implement it inside NDS. The rules above do configurable hard limitting and QoS among users first, and then do flow based QoS separate for each user, inside each user's limits. It's like having an independent and simplified sqm-scripts for each of your users.

If you just want an interface centric and shared flow based QoS, then qos-scripts or sqm-scripts would probably do a better job.

mwarning commented:
@drwyrm let us know when you think it is ready to be merged.
I may have some questions for you (have configure options or dependencies changed etc.), so we can update the documentation.

drwyrm commented:
@mwarning Until we have a way to set per user limits, there isn't much point in merging it, previously this was provided by binvoucher, but since libhttpd was replaced by libmicrohttpd there is no way to actually use most of the TrafficControl code.

I will write something to maybe use ndsctl to set per user limits or add some other way to set this up. I would be happy to implement binvoucher on current master, but there is already another pull request to do it, and I don't want to step over other people work.

About config options and dependencies: The kernel modules needed are completely different, now we need ifb, sch_hfsc and sch_codel instead of imq and sch_htb. The config options that changed were undocumented, I removed "dowloadimq" and "uploadimq" and added "ifb" which takes an integer as argument to select the ifbX interface that NDS should use. Also... I suggest making "ifb" mandatory in case "TrafficControl" is enabled instead of using a default value of 0, as it will conflict with all current qos implementations on OpenWRT otherwise.

Offtopic: What is your opinion on using a passwd like file for usernames, passwords and limits instead of allowing only a single user/pass site-wide? Wouldn't be hard to implement and would add a lot of value in my opinion.

lynxis commented:
hi, I'm actual in holidays, I'll look into this in the first 2 weeks.

mwarning commented:
@drwyrm yup, let's wait for the binvoucher merge request.

As for a username/pasword file, maybe it should be implemented as part of the binvoucher support.
Essentially call a script that can get passed a username/password in the environment variables.
Right now, we pass all stuff as arguments (mac, token, ..), but that becomes messy to handle at some point..

drwyrm commented:
Sounds good. Using an external script/executable to handle
username/password would also allow any other kind of authentication without
adding extra dependencies to NDS code, from http auth using external server
to radius, etc.. I also like the idea to pass arguments using env variables
instead of the command line. I will work on it as soon as possible. :)

2017-01-02 12:51 GMT-02:00 Moritz Warning <notifications@github.com>:

> @drwyrm <https://github.com/drwyrm> yup, let's wait for the binvoucher
> merge request.
>
> As for a username/pasword file, maybe it should be implemented as part of
> the binvoucher support.
> Essentially call a script that can get passed a username/password in the
> environment variables.
> Right now, we pass all stuff as arguments (mac, token, ..), but that
> becomes messy to handle at some point..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodogsplash/nodogsplash/pull/155#issuecomment-269981918>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAhRBrteBWhAkWya7h_h8e2iYphVpgPSks5rOQ7cgaJpZM4LWvjM>
> .
>



-- 
Daniel Ribeiro

mwarning commented:
@drwyrm yay :-)

porjo commented:
Any progress on this?

mwarning commented:
I think this PR got forgotten. :/
Let's see what can be done.

mwarning commented:
ok, merged since the previous code is usable anyway.
What was bugging me is that I don't know it work and never had the time to properly review it.
